### PR TITLE
New forest utilities

### PR DESCRIPTION
**Pull request by chanicpanic**
This improves the `ForestVisitor` and adds other Forest utilities. See #673.

**Summary of Changes**

- Small improvements and updates have been made to the existing `ForestVisitor` class. The most significant ones being the removal of a visit return value and improved handling of cycles.
- A new class, `ForestTransformer`, has been added as an extension of `ForestVisitor` to be the base class for bottom-up forest transformations.
- A specialized `ForestTransformer`, `ForestToParseTree`, replaces the old `ForestToTreeVisitor`, `ForestToAmbiguousTreeVisitor`, and `CompleteForestToAmbiguousTreeVisitor`.
- `TreeForestTransformer` has been written as a `ForestTransformer` with  an interface similar to the existing tree `Transformer`. The major difference is additional methods for ambiguity handling. While designed for controlling forest to tree transformations, clever subclass implementations may be able to perform arbitrary transformations.

### CODE REVIEW

File: lark/parsers/earley_forest.py [hunk 1]
```diff
@@ -208,25 +234,130 @@ def visit(self, root):
...
+
+    def __init__(self):
+        # results of transformations
+        self.data = dict()
+        # used to track parent nodes
+        self.node_stack = deque()
+
+    def transform(self, root):
+        """Perform a transformation on a Forest."""
+        self.node_stack.append('result')
```
Comment (erezsh) [inline lark/parsers/earley_forest.py:296]: What is the advantage of having `data['result']` rather than just `self.result`?

Comment (chanicpanic) [reply lark/parsers/earley_forest.py:296]: In `visit*node_out`, transformations are appended to the list of data for the parent node. By using `data['result']` and pushing `'result'` onto `node_stack`, I don't have to check whether the stack is empty in all those methods (whether the node is `root`).

File: lark/parsers/earley_forest.py [hunk 2]
```diff
@@ -208,25 +234,130 @@ def visit(self, root):
...
+
+    def visit_packed_node_in(self, node):
+        self.node_stack.append(id(node))
+        self.data[id(node)] = []
+        return node.children
+
+    def visit_token_node(self, node):
+        try:
+            self.data[self.node_stack[-1]].append(self.transform_token_node(node))
+        except Discard:
```
[the remark this answers is not in the document]
Comment (chanicpanic) [reply lark/parsers/earley_forest.py:328]: Do you mean put the transformation attempt under `try` and the append under `else`?

**Comment by chanicpanic** [comment]
A new PR with a cleaner history has been opened. #681

**Pull request closed without merging**